### ISSUE TITLE
high ram/memory usage

### ISSUE THREAD

**Issue by BO41** [eslint/eslint#12482]
**Tell us about your environment**
Environment Info:

Node version: v12.11.1
npm version: v6.12.0
Local ESLint version: v6.5.1
Global ESLint version: v6.5.1 (Currently used)

```bash
eslint 'src/**/*.ts'
```

**Are you willing to submit a pull request to fix this bug?**
depends on the scope

- Issue Type: `Bug`
- OS Version: `Linux x64 5.3.6-1-MANJARO`
- "@typescript-eslint/eslint-plugin": "2.4.0",
- "@typescript-eslint/parser": "2.4.0",
- "eslint": "6.5.1",
- "eslint-config-prettier": "6.4.0",
- "eslint-plugin-prettier": "3.1.1",
- "eslint-plugin-promise": "4.2.1",

The Ram usage of eslint is very high (~2,4GiB).
Is this normal for this many rules and plugins?
Is this a eslint or a typescript-eslint problem?

![ram](https://user-images.githubusercontent.com/29450609/67094508-4aa14100-f1b4-11e9-8cfc-3ca77944a73c.png)

<details>
<summary>eslintrc.json</summary>

```json
{
    "extends": [
        "eslint:recommended",
        "plugin:@typescript-eslint/eslint-recommended",
        "plugin:@typescript-eslint/recommended",
        "plugin:@typescript-eslint/recommended-requiring-type-checking",
        "prettier",
        "prettier/@typescript-eslint",
        "plugin:prettier/recommended",
        "plugin:promise/recommended"
    ],
    "parser": "@typescript-eslint/parser",
    "plugins": ["@typescript-eslint", "prettier", "promise"],
    "env": {
        "node": true
    },
    "parserOptions": {
        "ecmaVersion": 6,
        "sourceType": "module",
        "project": "./tsconfig.json",
        "ecmaFeatures": {
            "modules": true
        }
    },
    "reportUnusedDisableDirectives": true,
    "rules": {
        "prettier/prettier": "warn",
        "semi": ["warn", "always"],
        "prefer-const": "warn",
        "max-lines": [
            "error",
            {
                "max": 300,
                "skipComments": false,
                "skipBlankLines": false
            }
        ],
        "arrow-body-style": "error",
        "curly": "warn",
        "dot-notation": "error",
        "eol-last": "error",
        "no-caller": "error",
        "no-debugger": "error",
        "no-new-wrappers": "error",
        "no-throw-literal": "error",
        "no-undef-init": "error",
        "no-var": "error",
        "radix": "error",
        "no-alert": "error",
        "no-console": ["error", { "allow": ["warn", "error", "log"] }],
        "no-proto": "error",
        "no-new-func": "error",
        "lines-between-class-members": [
            "warn",
            "always",
            {
                "exceptAfterSingleLine": true
            }
        ],
        "spaced-comment": "warn",
        "guard-for-in": "warn",
        "prefer-spread": "off",
        "no-restricted-imports": ["error", "rxjs/Rx"],
        "quotes": ["warn", "single"],
        "quote-props": ["warn", "as-needed"],
        "max-lines-per-function": ["warn", { "max": 75, "skipComments": true }],
        "@typescript-eslint/no-namespace": "off",
        "@typescript-eslint/camelcase": "warn",
        "@typescript-eslint/type-annotation-spacing": "warn",
        "@typescript-eslint/no-inferrable-types": [
            "warn",
            {
                "ignoreParameters": true,
                "ignoreProperties": false
            }
        ],
        "@typescript-eslint/no-explicit-any": "warn",
        "@typescript-eslint/consistent-type-assertions": "warn",
        "@typescript-eslint/no-empty-interface": "error",
        "@typescript-eslint/no-misused-new": "error",
        "@typescript-eslint/no-non-null-assertion": "error",
        "@typescript-eslint/member-ordering": "warn",
        "@typescript-eslint/unified-signatures": "warn",
        "@typescript-eslint/consistent-type-definitions": ["warn", "interface"],
        "@typescript-eslint/class-name-casing": "warn",
        "@typescript-eslint/no-use-before-define": "off",
        "@typescript-eslint/unbound-method": [
            "error",
            {
                "ignoreStatic": true
            }
        ],
        "@typescript-eslint/explicit-member-accessibility": ["warn", { "accessibility": "no-public" }],
        "@typescript-eslint/explicit-function-return-type": ["warn", { "allowExpressions": true }]
    }
}
```
</details>

cc https://github.com/microsoft/vscode-eslint/issues/782

**Comment by g-plane** [comment]
Thanks for this issue.

I think there are several reasons:
- You're running `vscode-eslint` instead of ESLint itself when measuring RAM usage. `vscode-eslint` is a VS Code extension, and it's running a "server" at the background, which also costs memory.
- You're linting TypeScript files and enabling "project" option. If this option is enabled, `@typescript-eslint/parser` will invoke TypeScript compiler to retrieve type information. This step **_may_ heavily consume your CPU and/or RAM resource**.
- The `eslint-plugin-prettier` will invoke Prettier.

Possible solutions:
- Lint your files at CLI only with ESLint.
- Run prettier separately at CLI. Don't run Prettier as an ESLint plugin.

**Comment by BO41** [comment]
Thanks for the answer :)

>     * You're linting TypeScript files and enabling "project" option. If this option is enabled, `@typescript-eslint/parser` will invoke TypeScript compiler to retrieve type information. This step **_may_ heavily consume your CPU and/or RAM resource**.

What do you mean with that?
Wouldn't I lose much functionality when not having type information?

>     * Run prettier separately at CLI. Don't run Prettier as an ESLint plugin.

There isn't a way to still benefit from this integration for example so that prettier knows where and if to insert brackets.
---
Is there a way to lint only the current file (the file opened by VSCode) and files needed for type information?

**Comment by BO41** [comment]
so, cutting prettier integration doesn't change ram usage at all :/

**Comment by g-plane** [comment]
Don't measure your RAM usage with "vscode-eslint".

**Comment by BO41** [comment]
I won't use eslint regularly if I can't integrate it into my IDE.

BUT, even in the cli the ram usage doesn't change by removing prettier integration.

---

```
✖ 1564 problems (600 errors, 964 warnings)
  9 errors and 66 warnings potentially fixable with the `--fix` option.
```
Shouldn't eslint terminate after this output ↑?
It prints all warnings and errors and then hangs there filling the ram.
Seems to me like a memory leak.

**Comment by kaicataldo** [comment]
There are a lot of moving parts here and it's going to be difficult for to figure this out without having a minimal reproduction case. Any chance you can find the minimal reproduction steps needed to cause this (using ESLint directly, not an editor integration)?

**Comment by BO41** [comment]
typescript-eslint [released version 2.5.0](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v2.5.0) which drops the ram usage to about ~1GiB :tada: and also fixes the bug described above, where eslint wouldn't terminate (https://github.com/typescript-eslint/typescript-eslint/issues/1079) :tada: :tada: 

**Thank you eslint** team for the help.
A lower ram consumption is always welcome, but for now I am happy again :smiley:
